concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/LifoBlockingLimiter.java

```diff
@@ -179,6 +179,11 @@ private Optional<Listener> tryAcquire(ContextT context) {
                 synchronized (lock) {
                     backlog.removeLastOccurrence(event);
                 }
+                // if we acquired a token just as we were timing out then return it, otherwise the
+                // token would get lost
+                if (event.listener != null) {
+                    return event.listener;
+                }
                 return Optional.empty();
```

Comment:
```suggestion
               return event.listener;
```

Reply:
That suggestion doesn't look right to me. That would make this method return `null`, which it never should. I expect that would lead to a NullPointerException.

Reply:
Its Optional not null.

Reply:
it currently is a field of type Optional, that can also be null. that was also surprising to me.

to be exact: if the delegate has not been called yet then it was null. if the delegate has been called but its limit was exceeded then it was Optional.empty().

I just pushed a commit that initializes it to Optional.empty() to fix that awkwardness.